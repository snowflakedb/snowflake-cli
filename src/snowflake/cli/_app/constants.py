# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import Literal

PARAM_APPLICATION_NAME: Literal["snowcli"] = "snowcli"

# This is also defined on server side. Changing this parameter would require
# a change in https://github.com/snowflakedb/snowflake
INTERNAL_APPLICATION_NAME: Literal["SNOWFLAKE_CLI"] = "SNOWFLAKE_CLI"

# Authenticator types
AUTHENTICATOR_WORKLOAD_IDENTITY: Literal["WORKLOAD_IDENTITY"] = "WORKLOAD_IDENTITY"
AUTHENTICATOR_SNOWFLAKE_JWT: Literal["SNOWFLAKE_JWT"] = "SNOWFLAKE_JWT"
AUTHENTICATOR_USERNAME_PASSWORD_MFA: Literal[
    "username_password_mfa"
] = "username_password_mfa"
AUTHENTICATOR_OAUTH_AUTHORIZATION_CODE: Literal[
    "OAUTH_AUTHORIZATION_CODE"
] = "OAUTH_AUTHORIZATION_CODE"
