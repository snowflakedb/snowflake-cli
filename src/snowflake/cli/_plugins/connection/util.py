# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import json
import logging
import os
from enum import Enum
from functools import lru_cache
from typing import Any, Dict, Optional

from click.exceptions import ClickException
from snowflake.connector import SnowflakeConnection
from snowflake.connector.cursor import DictCursor

log = logging.getLogger(__name__)


ALLOWLIST_QUERY = "SELECT SYSTEM$ALLOWLIST()"
SNOWFLAKE_DEPLOYMENT = "SNOWFLAKE_DEPLOYMENT"
LOCAL_DEPLOYMENT_REGION: str = "us-west-2"


class MissingConnectionAccountError(ClickException):
    def __init__(self, conn: SnowflakeConnection):
        super().__init__(
            "Could not determine account by system call, configured account name, or configured host. Connection: "
            + repr(conn)
        )


class MissingConnectionRegionError(ClickException):
    def __init__(self, host: str | None):
        super().__init__(
            f"The connection host ({host}) was missing or not in "
            "the expected format "
            "(<account>.<deployment>.snowflakecomputing.com)"
        )


class UIParameter(Enum):
    NA_ENABLE_REGIONLESS_REDIRECT = "UI_SNOWSIGHT_ENABLE_REGIONLESS_REDIRECT"
    NA_EVENT_SHARING_V2 = "ENABLE_EVENT_SHARING_V2_IN_THE_SAME_ACCOUNT"
    NA_ENFORCE_MANDATORY_FILTERS = (
        "ENFORCE_MANDATORY_FILTERS_FOR_SAME_ACCOUNT_INSTALLATION"
    )
    NA_FEATURE_RELEASE_CHANNELS = "FEATURE_RELEASE_CHANNELS"


def get_ui_parameter(
    conn: SnowflakeConnection, parameter: UIParameter, default: Any
) -> Any:
    """
    Returns the value of a single UI parameter.
    If the parameter is not found, the default value is returned.
    """

    ui_parameters = get_ui_parameters(conn)
    return ui_parameters.get(parameter, default)


@lru_cache()
def get_ui_parameters(conn: SnowflakeConnection) -> Dict[UIParameter, Any]:
    """
    Returns the UI parameters from the SYSTEM$BOOTSTRAP_DATA_REQUEST function
    """

    parameters_to_fetch = [param.value for param in UIParameter]

    # Parsing of the Json and the filtering is happening here in Snowflake CLI
    # in order to avoid requiring a warehouse in Snowflake
    query = "call system$bootstrap_data_request('CLIENT_PARAMS_INFO')"
    *_, cursor = conn.execute_string(query)

    json_map = json.loads(cursor.fetchone()[0])

    return {
        UIParameter(row["name"]): row["value"]
        for row in json_map["clientParamsInfo"]
        if row["name"] in parameters_to_fetch
    }


def clear_ui_parameters_cache() -> None:
    """
    Clear the LRU cache for UI parameters to ensure fresh state.

    This function should be called between tests or operations to prevent
    state pollution from cached UI parameters. It clears the cache for
    get_ui_parameters function.
    """
    get_ui_parameters.cache_clear()


def is_regionless_redirect(conn: SnowflakeConnection) -> bool:
    """
    Determines if the deployment this connection refers to uses
    regionless URLs in Snowsight (/orgname/account) or regional URLs
    (/region/account). If we cannot determine the correct value we
    assume it's regionless, as this is true for most production deployments.
    """
    try:
        return get_ui_parameter(conn, UIParameter.NA_ENABLE_REGIONLESS_REDIRECT, True)
    except:
        log.warning(
            "Cannot determine regionless redirect; assuming True.", exc_info=True
        )
        return True


def get_host_region(host: str) -> str | None:
    """
    Looks for hosts of form
    <account>.[x.y.z].snowflakecomputing.com
    Returns the three-part [region identifier] or None.
    """
    host_parts = host.split(".")
    if host_parts[-1] == "local":
        return LOCAL_DEPLOYMENT_REGION
    elif len(host_parts) == 6:
        return ".".join(host_parts[1:4])
    return None


def guess_regioned_host_from_allowlist(conn: SnowflakeConnection) -> str | None:
    """
    Use SYSTEM$ALLOWLIST to find a regioned host (<account>.x.y.z.snowflakecomputing.com)
    that corresponds to the given Snowflake connection object.
    """
    try:
        *_, cursor = conn.execute_string(ALLOWLIST_QUERY, cursor_class=DictCursor)
        allowlist_tuples = json.loads(cursor.fetchone()["SYSTEM$ALLOWLIST()"])
        for t in allowlist_tuples:
            if t["type"] == SNOWFLAKE_DEPLOYMENT:
                if get_host_region(t["host"]) is not None:
                    return t["host"]
    except:
        log.warning(
            "Could not call SYSTEM$ALLOWLIST; returning an empty guess.", exc_info=True
        )
    return None


def get_region(conn: SnowflakeConnection) -> str:
    """
    Get the region of the given connection, or raise MissingConnectionRegionError.
    """
    if conn.host:
        if region := get_host_region(conn.host):
            return region

    if host := guess_regioned_host_from_allowlist(conn):
        if region := get_host_region(host):
            return region

    raise MissingConnectionRegionError(host or conn.host)


def get_context(conn: SnowflakeConnection) -> str:
    """
    Determines the first part of the path in a Snowsight URL.
    This could be a region or it could be an organization, depending
    on whether or not the underlying deployment uses regionless URLs.
    """
    if is_regionless_redirect(conn):
        *_, cursor = conn.execute_string(
            f"select system$return_current_org_name()", cursor_class=DictCursor
        )
        return cursor.fetchone()["SYSTEM$RETURN_CURRENT_ORG_NAME()"]

    return get_region(conn)


def get_account(conn: SnowflakeConnection) -> str:
    """
    Determines the account that this connection refers to.
    """
    try:
        *_, cursor = conn.execute_string(
            f"select current_account_name()", cursor_class=DictCursor
        )
        return cursor.fetchone()["CURRENT_ACCOUNT_NAME()"].lower()
    except Exception as e:
        # try to extract the account from the connection information
        if conn.account:
            return conn.account

        if conn.host:
            host_parts = conn.host.split(".")
            return host_parts[0]

        raise MissingConnectionAccountError(conn)


def get_snowsight_host(conn: SnowflakeConnection) -> str:
    try:
        *_, cursor = conn.execute_string(
            f"select system$get_snowsight_host()", cursor_class=DictCursor
        )
        return cursor.fetchone()["SYSTEM$GET_SNOWSIGHT_HOST()"]
    except Exception as e:
        # if we cannot determine the host, assume we're on prod
        return "https://app.snowflake.com"


def make_snowsight_url(conn: SnowflakeConnection, path: str) -> str:
    """
    Returns a URL on the correct Snowsight instance for the connected account.
    The path that is passed in must already be properly URL-encoded, and
    can optionally contain a hash/fragment (e.g. #).

    See also identifier_for_url.
    """
    snowsight_host = get_snowsight_host(conn)
    deployment = get_context(conn)
    account = get_account(conn)
    path_with_slash = path if path.startswith("/") else f"/{path}"
    return f"{snowsight_host}/{deployment}/{account}{path_with_slash}"


def strip_if_value_present(value: Optional[str]) -> Optional[str]:
    return value.strip() if value else value


def ensure_that_path_exist(path: Optional[str]) -> Optional[str]:
    if path and not os.path.exists(path):
        raise ClickException(f"Path {path} does not exist.")
    return path


def strip_and_check_if_exists(value: Optional[str]) -> Optional[str]:
    return ensure_that_path_exist(strip_if_value_present(value))
