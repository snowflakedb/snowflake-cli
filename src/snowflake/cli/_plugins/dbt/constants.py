# Copyright (c) 2025 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

RESULT_COLUMN_NAME = "SUCCESS"
OUTPUT_COLUMN_NAME = "STDOUT"

DBT_COMMANDS = [
    "build",
    "compile",
    "deps",
    "list",
    "parse",
    "run",
    "seed",
    "show",
    "snapshot",
    "test",
]

UNSUPPORTED_COMMANDS = [
    "clean",
    "clone",
    "debug",
    "docs",
    "init",
    "retry",
    "source",
]
