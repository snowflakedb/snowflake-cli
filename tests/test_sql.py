# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from pathlib import Path
from tempfile import NamedTemporaryFile, TemporaryDirectory
from unittest import mock

import pytest
from snowflake.cli._plugins.sql.snowsql_templating import transpile_snowsql_templates
from snowflake.cli.api.constants import ObjectType
from snowflake.cli.api.exceptions import (
    ShowSpecificObjectMultipleRowsError,
    SnowflakeSQLExecutionError,
)
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.project.util import identifier_to_show_like_pattern
from snowflake.cli.api.sql_execution import SqlExecutionMixin, VerboseCursor
from snowflake.connector.cursor import DictCursor

from tests.testing_utils.result_assertions import assert_that_result_is_usage_error


@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin._execute_string")
def test_sql_execute_query(mock_execute, runner, mock_cursor):
    mock_execute.return_value = (mock_cursor(["row"], []) for _ in range(1))

    result = runner.invoke(["sql", "-q", "query"])

    assert result.exit_code == 0
    mock_execute.assert_called_once_with("query", cursor_class=VerboseCursor)


@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin._execute_string")
def test_sql_execute_file(mock_execute, runner, mock_cursor, named_temporary_file):
    mock_execute.return_value = (mock_cursor(["row"], []) for _ in range(1))
    query = "query from file"

    with named_temporary_file() as tmp_file:
        tmp_file.write_text(query)
        result = runner.invoke(["sql", "-f", tmp_file])

    assert result.exit_code == 0
    mock_execute.assert_called_once_with(query, cursor_class=VerboseCursor)


@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin._execute_string")
def test_sql_execute_multiple_file(mock_execute, runner, mock_cursor):
    mock_execute.return_value = (mock_cursor(["row"], []) for _ in range(1))
    query = "query from file"

    with TemporaryDirectory() as tmp_dir:
        f1 = Path(tmp_dir).joinpath("f1.sql")
        f2 = Path(tmp_dir).joinpath("f2.sql")

        f1.write_text(query)
        f2.write_text(query)
        result = runner.invoke(["sql", "-f", f1, "-f", f2])

    assert result.exit_code == 0
    mock_execute.assert_has_calls(
        [
            mock.call(query, cursor_class=VerboseCursor),
            mock.call(query, cursor_class=VerboseCursor),
        ]
    )


@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin._execute_string")
def test_sql_execute_from_stdin(mock_execute, runner, mock_cursor):
    mock_execute.return_value = (mock_cursor(["row"], []) for _ in range(1))
    query = "query from input"

    result = runner.invoke(["sql", "-i"], input=query)

    assert result.exit_code == 0
    mock_execute.assert_called_once_with(query, cursor_class=VerboseCursor)


def test_sql_help_if_no_query_file_or_stdin(runner, os_agnostic_snapshot):
    result = runner.invoke(["sql"])
    assert result.exit_code == 0, result.output
    assert result.output == os_agnostic_snapshot


def test_sql_fails_if_query_and_stdin_and_file_provided(runner):
    with NamedTemporaryFile("r") as tmp_file:
        result = runner.invoke(["sql", "-i", "-q", "foo", "-f", tmp_file.name])
        assert_that_result_is_usage_error(
            result,
            f"Parameters '--query' and '--stdin' are incompatible and cannot be used simultaneously.",
        )


@pytest.mark.parametrize("inputs", [(("-i",), "stdin"), (("-q", "foo"), "query")])
def test_sql_fails_if_other_input_and_file_provided(runner, inputs):
    with NamedTemporaryFile("r") as tmp_file:
        result = runner.invoke(["sql", *(inputs[0]), "-f", tmp_file.name])
        assert_that_result_is_usage_error(
            result,
            f"Parameters '--filename' and '--{inputs[1]}' are incompatible and cannot be used simultaneously.",
        )


def test_sql_fails_if_query_and_stdin_provided(runner):
    result = runner.invoke(["sql", "-q", "fooo", "-i"])
    assert_that_result_is_usage_error(
        result,
        "Parameters '--stdin' and '--query' are incompatible and cannot be used simultaneously. ",
    )


@mock.patch("snowflake.cli._app.snow_connector.connect_to_snowflake")
def test_sql_overrides_connection_configuration(mock_conn, runner, mock_cursor):
    mock_conn.return_value.execute_string.return_value = [mock_cursor(["row"], [])]

    result = runner.invoke(
        [
            "sql",
            "-q",
            "select 1",
            "--connection",
            "connectionName",
            "--accountname",
            "accountnameValue",
            "--username",
            "usernameValue",
            "--dbname",
            "dbnameValue",
            "--schemaname",
            "schemanameValue",
            "--rolename",
            "rolenameValue",
            "--warehouse",
            "warehouseValue",
            "--password",
            "passFromTest",
            "--diag-log-path",
            "/tmp",
        ],
        catch_exceptions=False,
    )

    assert result.exit_code == 0, result.output
    mock_conn.assert_called_once_with(
        temporary_connection=False,
        enable_diag=False,
        diag_log_path=Path("/tmp"),
        connection_name="connectionName",
        account="accountnameValue",
        user="usernameValue",
        password="passFromTest",
        database="dbnameValue",
        schema="schemanameValue",
        role="rolenameValue",
        warehouse="warehouseValue",
    )


@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin.execute_query")
def test_show_specific_object(mock_execute, mock_cursor):
    mock_columns = ["id", "created_on"]
    mock_row_dict = {c: r for c, r in zip(mock_columns, ["EXAMPLE_ID", "dummy"])}
    cursor = mock_cursor(rows=[mock_row_dict], columns=mock_columns)
    mock_execute.return_value = cursor
    result = SqlExecutionMixin().show_specific_object(
        "objects", "example_id", name_col="id"
    )
    mock_execute.assert_called_once_with(
        r"show objects like 'EXAMPLE\\_ID'", cursor_class=DictCursor
    )
    assert result == mock_row_dict


@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin.execute_query")
def test_show_specific_object_in_clause(mock_execute, mock_cursor):
    mock_columns = ["name", "created_on"]
    mock_row_dict = {c: r for c, r in zip(mock_columns, ["AbcDef", "dummy"])}
    cursor = mock_cursor(rows=[mock_row_dict], columns=mock_columns)
    mock_execute.return_value = cursor
    result = SqlExecutionMixin().show_specific_object(
        "objects", '"AbcDef"', in_clause="in database mydb"
    )
    mock_execute.assert_called_once_with(
        r"show objects like 'AbcDef' in database mydb", cursor_class=DictCursor
    )
    assert result == mock_row_dict


@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin.execute_query")
def test_show_specific_object_no_match(mock_execute, mock_cursor):
    mock_columns = ["id", "created_on"]
    mock_row_dict = {c: r for c, r in zip(mock_columns, ["OTHER_ID", "dummy"])}
    cursor = mock_cursor(rows=[mock_row_dict], columns=mock_columns)
    mock_execute.return_value = cursor
    result = SqlExecutionMixin().show_specific_object(
        "objects", "example_id", name_col="id"
    )
    mock_execute.assert_called_once_with(
        r"show objects like 'EXAMPLE\\_ID'", cursor_class=DictCursor
    )
    assert result is None


@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin.execute_query")
def test_show_specific_object_sql_execution_error(mock_execute):
    cursor = mock.Mock(spec=DictCursor)
    cursor.rowcount = None
    mock_execute.return_value = cursor
    with pytest.raises(SnowflakeSQLExecutionError):
        SqlExecutionMixin().show_specific_object("objects", "example_id", name_col="id")
    mock_execute.assert_called_once_with(
        r"show objects like 'EXAMPLE\\_ID'", cursor_class=DictCursor
    )


@pytest.mark.parametrize(
    "fqn, expected_in_clause",
    [
        (
            FQN(None, None, "func(number, number)"),
            None,
        ),
        (FQN(None, None, "name"), None),
        (FQN(None, "schema", "name"), "in schema schema"),
        (FQN("db", "schema", "name"), "in schema db.schema"),
    ],
)
def test_qualified_name_to_in_clause(fqn, expected_in_clause):
    assert SqlExecutionMixin._qualified_name_to_in_clause(fqn) == (  # noqa: SLF001
        fqn.name,
        expected_in_clause,
    )


@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin.execute_query")
@mock.patch(
    "snowflake.cli.api.sql_execution.SqlExecutionMixin._qualified_name_to_in_clause"
)
def test_show_specific_object_qualified_name(
    mock_qualified_name_to_in_clause, mock_execute_query, mock_cursor
):
    name = "db.schema.obj"
    unqualified_name = "obj"
    name_in_clause = "in schema db.schema"
    mock_columns = ["name", "created_on"]
    mock_row_dict = {c: r for c, r in zip(mock_columns, [unqualified_name, "date"])}
    cursor = mock_cursor(rows=[mock_row_dict], columns=mock_columns)
    mock_execute_query.return_value = cursor

    mock_qualified_name_to_in_clause.return_value = (unqualified_name, name_in_clause)
    SqlExecutionMixin().show_specific_object("objects", name)
    mock_execute_query.assert_called_once_with(
        f"show objects like {identifier_to_show_like_pattern(unqualified_name)} {name_in_clause}",
        cursor_class=DictCursor,
    )


@mock.patch(
    "snowflake.cli.api.sql_execution.SqlExecutionMixin._qualified_name_to_in_clause"
)
def test_show_specific_object_qualified_name_and_in_clause_error(
    mock_qualified_name_to_in_clause,
):
    object_name = "db.schema.name"
    mock_qualified_name_to_in_clause.return_value = ("name", "in schema db.schema")
    with pytest.raises(SqlExecutionMixin.InClauseWithQualifiedNameError):
        SqlExecutionMixin().show_specific_object(
            "objects", object_name, in_clause="in database db"
        )
    mock_qualified_name_to_in_clause.assert_called_once_with(
        FQN.from_string(object_name)
    )


@mock.patch("snowflake.cli.api.sql_execution.SqlExecutionMixin.execute_query")
def test_show_specific_object_multiple_rows(mock_execute_query):
    cursor = mock.Mock(spec=DictCursor)
    cursor.rowcount = 2
    mock_execute_query.return_value = cursor
    with pytest.raises(ShowSpecificObjectMultipleRowsError) as err:
        SqlExecutionMixin().show_specific_object("objects", "name", name_col="id")
    assert err.match("Received multiple rows")
    mock_execute_query.assert_called_once_with(
        r"show objects like 'NAME'", cursor_class=DictCursor
    )


@pytest.mark.parametrize(
    "_object",
    [
        ObjectType.WAREHOUSE,
        ObjectType.ROLE,
        ObjectType.DATABASE,
    ],
)
@mock.patch("snowflake.cli.api.sql_execution.SqlExecutionMixin.execute_query")
def test_use_command(mock_execute_query, _object):
    SqlExecutionMixin().use(object_type=_object, name="foo_name")
    mock_execute_query.assert_called_once_with(f"use {_object.value.sf_name} foo_name")


@pytest.mark.parametrize(
    "query",
    [
        "select &{ aaa }.&{ bbb }",
        "select &aaa.&bbb",
        "select &aaa.&{ bbb }",
        "select <% aaa %>.<% bbb %>",
    ],
)
@mock.patch("snowflake.cli._plugins.sql.commands.SqlManager._execute_string")
def test_rendering_of_sql(mock_execute_query, query, runner):
    result = runner.invoke(["sql", "-q", query, "-D", "aaa=foo", "-D", "bbb=bar"])
    assert result.exit_code == 0, result.output
    mock_execute_query.assert_called_once_with(
        "select foo.bar", cursor_class=VerboseCursor
    )


@mock.patch("snowflake.cli._plugins.sql.commands.SqlManager._execute_string")
def test_old_template_syntax_causes_warning(mock_execute_query, runner):
    result = runner.invoke(["sql", "-q", "select &{ aaa }", "-D", "aaa=foo"])
    assert result.exit_code == 0
    assert (
        "Warning: &{ ... } syntax is deprecated. Use <% ... %> syntax instead."
        in result.output
    )
    mock_execute_query.assert_called_once_with("select foo", cursor_class=VerboseCursor)


@mock.patch("snowflake.cli._plugins.sql.commands.SqlManager._execute_string")
def test_mixed_template_syntax_error(mock_execute_query, runner):
    result = runner.invoke(
        ["sql", "-q", "select <% aaa %>.&{ bbb }", "-D", "aaa=foo", "-D", "bbb=bar"]
    )
    assert result.exit_code == 1
    assert "The SQL query mixes &{ ... } syntax and <% ... %> syntax." in result.output


@pytest.mark.parametrize(
    "query", ["select &{ foo }", "select &foo", "select <% foo %>"]
)
def test_execution_fails_if_unknown_variable(runner, query):
    result = runner.invoke(["sql", "-q", query, "-D", "bbb=1"])
    assert "SQL template rendering error: 'foo' is undefined" in result.output


@pytest.mark.parametrize(
    "text, expected",
    [
        # Test escaping
        ("&&foo", "&foo"),
        ("select *  from &&foo join bar", "select *  from &foo join bar"),
        # Test basic usage
        ("&foo", "&{ foo }"),
        ("select *  from &foo join bar", "select *  from &{ foo } join bar"),
        # Test templating is ignored
        ("&{ foo }", "&{ foo }"),
        ("select *  from &{ foo } join bar", "select *  from &{ foo } join bar"),
    ],
)
def test_snowsql_compatibility(text, expected):
    assert transpile_snowsql_templates(text) == expected


@pytest.mark.parametrize("template_start,template_end", [("&{", "}"), ("<%", "%>")])
@mock.patch("snowflake.cli._plugins.sql.commands.SqlManager._execute_string")
def test_uses_variables_from_snowflake_yml(
    mock_execute_query, project_directory, runner, template_start, template_end
):
    with project_directory("sql_templating"):
        result = runner.invoke(
            ["sql", "-q", f"select {template_start} ctx.env.sf_var {template_end}"]
        )

    assert result.exit_code == 0
    mock_execute_query.assert_called_once_with(
        "select foo_value", cursor_class=VerboseCursor
    )


@pytest.mark.parametrize("template_start,template_end", [("&{", "}"), ("<%", "%>")])
@mock.patch("snowflake.cli._plugins.sql.commands.SqlManager._execute_string")
def test_uses_variables_from_snowflake_local_yml(
    mock_execute_query, project_directory, runner, template_start, template_end
):
    with project_directory("sql_templating"):
        result = runner.invoke(
            [
                "sql",
                "-q",
                f"select {template_start} ctx.env.sf_var_override {template_end}",
            ]
        )

    assert result.exit_code == 0
    mock_execute_query.assert_called_once_with(
        "select foo_value_override", cursor_class=VerboseCursor
    )


@pytest.mark.parametrize("template_start,template_end", [("&{", "}"), ("<%", "%>")])
@mock.patch("snowflake.cli._plugins.sql.commands.SqlManager._execute_string")
def test_uses_variables_from_cli_are_added_outside_context(
    mock_execute_query, project_directory, runner, template_start, template_end
):
    with project_directory("sql_templating"):
        result = runner.invoke(
            [
                "sql",
                "-q",
                f"select {template_start} ctx.env.sf_var {template_end} {template_start} other {template_end}",
                "-D",
                "other=other_value",
            ]
        )

    assert result.exit_code == 0, result.output
    mock_execute_query.assert_called_once_with(
        "select foo_value other_value", cursor_class=VerboseCursor
    )


@pytest.mark.parametrize(
    "option,expected",
    [
        ("--retain-comments", "SELECT 42;\n-- Commented line\n    SELECT 1;\n"),
        ("", "SELECT 42;\nSELECT 1;\n"),
    ],
)
@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin._execute_string")
def test_comments_are_handled_correctly_from_file(
    mock_execute, runner, mock_cursor, named_temporary_file, option, expected
):
    mock_execute.return_value = (mock_cursor(["row"], []) for _ in range(1))
    query = """
    SELECT 42;
    -- Commented line
    SELECT 1;
    """

    arguments = ["sql", "-f", query]
    if option:
        arguments.append(option)

    with named_temporary_file() as tmp_file:
        tmp_file.write_text(query)
        arguments = ["sql", "-f", tmp_file]
        if option:
            arguments.append(option)

        result = runner.invoke(arguments)

    assert result.exit_code == 0
    mock_execute.assert_called_once_with(expected, cursor_class=VerboseCursor)


@pytest.mark.parametrize(
    "option,expected",
    [
        (
            "--retain-comments",
            "SELECT 42;\n-- Commented line\n    SELECT 1;\n--another comment;",
        ),
        ("", "SELECT 42;\nSELECT 1;\n"),
    ],
)
@mock.patch("snowflake.cli._plugins.sql.manager.SqlExecutionMixin._execute_string")
def test_comments_are_handled_correctly_from_query(
    mock_execute, runner, mock_cursor, named_temporary_file, option, expected
):
    mock_execute.return_value = (mock_cursor(["row"], []) for _ in range(1))
    query = """
    SELECT 42;
    -- Commented line
    SELECT 1;
    --another comment;
    """

    arguments = ["sql", "-q", query]
    if option:
        arguments.append(option)

    result = runner.invoke(arguments)

    assert result.exit_code == 0
    mock_execute.assert_called_once_with(expected, cursor_class=VerboseCursor)
