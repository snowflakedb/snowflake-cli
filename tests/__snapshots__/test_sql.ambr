# serializer version: 1
# name: test_sql_help_if_no_query_file_or_stdin
  '''
                                                                                  
   Usage: root sql [OPTIONS]                                                      
                                                                                  
   Executes Snowflake query.                                                      
   Use either query, filename or input option. Query to execute can be specified  
   using query option, filename option (all queries from file will be executed)   
   or via stdin by piping output from other command. For example cat my.sql |     
   snow sql -i. The command supports variable substitution that happens on        
   client-side.                                                                   
                                                                                  
  +- Options --------------------------------------------------------------------+
  | --query            -q      TEXT  Query to execute.                           |
  | --filename         -f      FILE  File to execute.                            |
  | --stdin            -i            Read the query from standard input. Use it  |
  |                                  when piping input to this command.          |
  | --variable         -D      TEXT  String in format of key=value. If provided  |
  |                                  the SQL content will be treated as template |
  |                                  and rendered using provided data.           |
  | --retain-comments                Retains comments in queries passed to       |
  |                                  Snowflake                                   |
  | --project          -p      TEXT  Path where the Snowflake project is stored. |
  |                                  Defaults to the current working directory.  |
  | --env                      TEXT  String in the format key=value. Overrides   |
  |                                  variables from the env section used for     |
  |                                  templates.                                  |
  | --help             -h            Show this message and exit.                 |
  +------------------------------------------------------------------------------+
  +- Connection configuration ---------------------------------------------------+
  | --connection,--environment     -c      TEXT     Name of the connection, as   |
  |                                                 defined in your config.toml  |
  |                                                 file. Default: default.      |
  | --host                                 TEXT     Host address for the         |
  |                                                 connection. Overrides the    |
  |                                                 value specified for the      |
  |                                                 connection.                  |
  | --port                                 INTEGER  Port for the connection.     |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --account,--accountname                TEXT     Name assigned to your        |
  |                                                 Snowflake account. Overrides |
  |                                                 the value specified for the  |
  |                                                 connection.                  |
  | --user,--username                      TEXT     Username to connect to       |
  |                                                 Snowflake. Overrides the     |
  |                                                 value specified for the      |
  |                                                 connection.                  |
  | --password                             TEXT     Snowflake password.          |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --authenticator                        TEXT     Snowflake authenticator.     |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --private-key-file,--privateâ€¦          TEXT     Snowflake private key file   |
  |                                                 path. Overrides the value    |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --token-file-path                      TEXT     Path to file with an OAuth   |
  |                                                 token that should be used    |
  |                                                 when connecting to Snowflake |
  | --database,--dbname                    TEXT     Database to use. Overrides   |
  |                                                 the value specified for the  |
  |                                                 connection.                  |
  | --schema,--schemaname                  TEXT     Database schema to use.      |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --role,--rolename                      TEXT     Role to use. Overrides the   |
  |                                                 value specified for the      |
  |                                                 connection.                  |
  | --warehouse                            TEXT     Warehouse to use. Overrides  |
  |                                                 the value specified for the  |
  |                                                 connection.                  |
  | --temporary-connection         -x               Uses a connection defined    |
  |                                                 with command line            |
  |                                                 parameters, instead of one   |
  |                                                 defined in config            |
  | --mfa-passcode                         TEXT     Token to use for             |
  |                                                 multi-factor authentication  |
  |                                                 (MFA)                        |
  | --enable-diag                                   Whether to generate a        |
  |                                                 connection diagnostic        |
  |                                                 report.                      |
  | --diag-log-path                        TEXT     Path for the generated       |
  |                                                 report. Defaults to system   |
  |                                                 temporary directory.         |
  | --diag-allowlist-path                  TEXT     Path to a JSON file          |
  |                                                 containing allowlist         |
  |                                                 parameters.                  |
  +------------------------------------------------------------------------------+
  +- Global configuration -------------------------------------------------------+
  | --format           [TABLE|JSON]  Specifies the output format.                |
  |                                  [default: TABLE]                            |
  | --verbose  -v                    Displays log entries for log levels info    |
  |                                  and higher.                                 |
  | --debug                          Displays log entries for log levels debug   |
  |                                  and higher; debug logs contain additional   |
  |                                  information.                                |
  | --silent                         Turns off intermediate output to console.   |
  +------------------------------------------------------------------------------+
  
  
  '''
# ---
